File: lintcode/cpp/Majority_Number.cpp
class Solution {
public:
    /**
     * @param nums: A list of integers
     * @return: The majority number
     */
    int majorityNumber(vector<int> nums) {
        // write your code here
        int cand;
        int cnt = 0;
        for(int i = 0; i < nums.size(); i++){
            if(cnt == 0){
                cand = nums[i];
                cnt = 1;
            }else{
                if(cand == nums[i]) cnt++;
                else cnt--;
            }
        }
        
        return cand;
    }
};

File: lintcode/java/MajorityNumber.java
public class Solution {
    /**
     * @param nums: a list of integers
     * @return: find a  majority number
     */
    public int majorityNumber(List<Integer> nums) {
        if(nums == null || nums.isEmpty()) {
            throw new IllegalArgumentException("xxx");
        }
        
        int num = 0;
        int count = 0;
        for(int x : nums) {
            if(count == 0) {
                num = x;
                count++;
                continue;
            }
            
            if(x == num) count++;
            else count--;
        }
        
        return num;
    }
}
